server/src/main/java/org/apache/druid/segment/realtime/appenderator/AppenderatorImpl.java

```diff
@@ -347,7 +348,8 @@ public AppenderatorAddResult add(
           }
         }
 
-        if (!skipBytesInMemoryOverheadCheck && bytesCurrentlyInMemory.get() - bytesToBePersisted > maxBytesTuningConfig) {
+        if (!skipBytesInMemoryOverheadCheck
```

Comment:
Hmm, I think the memory overhead calculation checks are going to end up triggering the supplier which will map the segment? Specifically `calculateMMappedHydrantMemoryInUsed`, which is going to try to get the storage adapter to count the number of columns. I think `FireHydrant` is going to need some way to track whether or not the segment has been mapped (related to my earlier thread on why the supplier might also be more suitable to live in `FireHydrant` somehow, so that we can have some side effect to let the hydrant know the mapping has happened)

Reply:
Good catch... I did not see memory being mapped in my tests but it was because I had "skipBytesInMemoryOverheadCheck": true in the ingestion spec....

Reply:
I refactored the memory calculations a little in my last commit, enough to make it work, but they still need more work (judging from my brief look at the heap dumps)